### PR TITLE
8305406: Add @spec tags in java.base/java.* (part 2)

### PR DESCRIPTION
**Pull request by hns**
Please review a doc update to add `@spec` tags to crypto and security APIs in `java.base`. 

This was authored and proposed as #13336 by @jonathan-gibbons as part of an effort to [add `@spec` tags and an external specifications page to API documentation][JDK-8296546]. The original PR was reviewed and approved, yet for some reason it was never integrated.

[JDK-8296546]: https://bugs.openjdk.org/browse/JDK-8296546

Since I couldn't reopen Jon's PR I created a new pull request based the original commits. They still apply cleanly, and I made sure all links are still valid and working. I added a commit to update the copyright header dates.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305406](https://bugs.openjdk.org/browse/JDK-8305406): Add @spec tags in java.base/java.* (part 2) (**Enhancement** - P3)

### Reviewers
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)

### Contributors
 * Daniel Jeliński `<djelinski@openjdk.org>`
 * Hannes Wallnöfer `<hannesw@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21326/head:pull/21326` \
`$ git checkout pull/21326`

Update a local copy of the PR: \
`$ git checkout pull/21326` \
`$ git pull https://git.openjdk.org/jdk.git pull/21326/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21326`

View PR using the GUI difftool: \
`$ git pr show -t 21326`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21326.diff">https://git.openjdk.org/jdk/pull/21326.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21326#issuecomment-2391526516)

### CODE REVIEW

**Comment by valeriepeng** [comment]
Is "https://tools.ietf.org/html/rfc8103" considered external spec? It is mentioned in com.sun.crypto.provider.ChaCha20Poly1305Parameters class but not covered in this PR. Is there any additional condition for an external reference to be included? How about an external reference to a section of a specification? E.g. http://www.w3.org/TR/xmlenc-core/#sec-Alg-SymmetricKeyWrap in com.sun.crypto.provider.DESedeWrapCipher class which is also not covered by this PR.

File: src/java.base/share/classes/java/security/Key.java [hunk 1]
```diff
@@ -86,6 +86,10 @@
  * Security Appendix</a>
  * of the <cite>Java Object Serialization Specification</cite> for more information.
  *
+ * @spec serialization/index.html Java Object Serialization Specification
```
Comment (valeriepeng) [inline src/java.base/share/classes/java/security/Key.java:89]: How about "Java Security Standard Algorithm Names Specification"? Should its reference be covered by @spec tag?

**Comment by jonathan-gibbons** [comment]
> Is "https://tools.ietf.org/html/rfc8103" considered external spec? It is mentioned in com.sun.crypto.provider.ChaCha20Poly1305Parameters class but not covered in this PR. Is there any additional condition for an external reference to be included? How about an external reference to a section of a specification? E.g. http://www.w3.org/TR/xmlenc-core/#sec-Alg-SymmetricKeyWrap in com.sun.crypto.provider.DESedeWrapCipher class which is also not covered by this PR.

Valerie,

The intent of `@spec` is to include any external specification that is referenced in the public API documentation. As a rule of thumb, an external specification is any one given by name and/or an absolute URL. Note, not all references with absolute URLs are specifications. Some may be references to additional non-normative sources such as guides and similar materials.

This PR, and others that have preceded it, are mechanically derived by examining `@see` and `{@link}` tags in public API documentation. Once this initial work has been completed, it will be up to individual component teams to update any `@spec` tags in their area, either adding new tags, or modifying/updating the URLs of existing tags.

File: src/java.base/share/classes/java/security/package-info.java [hunk 2]
```diff
@@ -79,6 +79,7 @@
  *
  * </ul>
  *
+ * @spec security/standard-names.html Java Security Standard Algorithm Names
```
Comment (valeriepeng) [inline src/java.base/share/classes/java/security/package-info.java:82]: How about the other 2 under `Package Specification` (from line 49-53)? Does this PR need to cover them also?

Comment (valeriepeng) [reply src/java.base/share/classes/java/security/package-info.java:82]: Same question applies for the java/security/interfaces/package-info.java file as well.

File: src/java.base/share/classes/java/security/cert/package-info.java [hunk 3]
```diff
@@ -52,6 +52,10 @@
  *   <li> {@extLink security_guide_pki Java PKI Programmer's Guide}
  * </ul>
  *
+ * @spec security/standard-names.html Java Security Standard Algorithm Names
+ * @spec https://www.rfc-editor.org/info/rfc5280
+ *      RFC 5280: Internet X.509 Public Key Infrastructure Certificate
+ *              and Certificate Revocation List (CRL) Profile
```
Comment (valeriepeng) [inline src/java.base/share/classes/java/security/cert/package-info.java:58]: How about RFC 2560 (line 39-40)?

**Review by hns — COMMENTED** [review]
Thanks for looking at this, Valerie. As Jon said in his comment, this PR is based on automated detection of @see and {@link} tags, and therefore not necessarily complete. 

That said I will add `@spec` for those cases you pointed out that I think are unambigous (see inline comments). Others that I think need more consideration and possibly updates of other parts of the doc comments I would prefer to be handled by the component maintainers.

**Comment by hns** [comment]
@valeriepeng I added two more `@spec` tags as suggested by your feedback (the ones I felt confident about).

I'm also uploading new API docs with the "External Specifications" page here: 
https://cr.openjdk.org/~hannesw/8305406/api.01/external-specs.html

File: src/java.base/share/classes/javax/net/ssl/SSLParameters.java [hunk 4]
```diff
@@ -300,6 +300,7 @@ public String getEndpointIdentificationAlgorithm() {
      *     Java Security Standard Algorithm Names</a> document
      *     for information about standard algorithm names.
      *
+     * @spec security/standard-names.html Java Security Standard Algorithm Names
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply src/java.base/share/classes/javax/net/ssl/SSLParameters.java:303]: Never mind, I see that the scope is only about absolute URLs in one of the earlier replies.

**Comment by valeriepeng** [comment]
> > Is "https://tools.ietf.org/html/rfc8103" considered external spec? It is mentioned in com.sun.crypto.provider.ChaCha20Poly1305Parameters class but not covered in this PR. Is there any additional condition for an external reference to be included? How about an external reference to a section of a specification? E.g. http://www.w3.org/TR/xmlenc-core/#sec-Alg-SymmetricKeyWrap in com.sun.crypto.provider.DESedeWrapCipher class which is also not covered by this PR.
> 
> Valerie,
> 
> The intent of `@spec` is to include any external specification that is referenced in the public API documentation. As a rule of thumb, an external specification is any one given by name and/or an absolute URL. Note, not all references with absolute URLs are specifications. Some may be references to additional non-normative sources such as guides and similar materials.
> 
> This PR, and others that have preceded it, are mechanically derived by examining `@see` and `{@link}` tags in public API documentation. Once this initial work has been completed, it will be up to individual component teams to update any `@spec` tags in their area, either adding new tags, or modifying/updating the URLs of existing tags.

I see. Thanks for the clarification.

**Review by valeriepeng — APPROVED** [review]

**Comment by hns** [comment]
Thanks!

/integrate

**Pull request closed without merging**